sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsContainer.java

```diff
@@ -0,0 +1,201 @@
+   */
+  public MetricsContainer(String stepName) {
+    this.stepName = stepName;
+  }
+
+  /**
+   * Set the {@link MetricsContainer} for the current thread.
+   */
+  public static void setMetricsContainer(MetricsContainer container) {
+    CONTAINER_FOR_THREAD.set(container);
+  }
+
+  /**
+   * Clear the {@link MetricsContainer} for the current thread.
+   */
+  public static void unsetMetricsContainer() {
+    CONTAINER_FOR_THREAD.remove();
+  }
+
+  /**
+   * Return the {@link MetricsContainer} for the current thread.
+   */
+  public static MetricsContainer getCurrentContainer() {
+    MetricsContainer container = CONTAINER_FOR_THREAD.get();
+    if (container == null) {
+      throw new IllegalStateException("Must call setMetricsContainer before reporting metrics.");
+    }
+    return container;
+  }
+
+  /**
+   * Return the {@link CounterCell} that should be used for implementing the given
+   * {@code metricName} in this container.
+   */
+  public CounterCell getOrCreateCounter(MetricName metricName) {
+    return counters.getOrCreate(metricName);
+  }
+
+  public DistributionCell getOrCreateDistribution(MetricName metricName) {
+    return distributions.getOrCreate(metricName);
+  }
+
+  private <UpdateT, CellT extends MetricCell<UpdateT>>
+  ImmutableList<MetricUpdate<UpdateT>> extractUpdates(
+      MetricsMap<MetricName, CellT> cells,
+      boolean includeZero) {
+    ImmutableList.Builder<MetricUpdate<UpdateT>> updates = ImmutableList.builder();
+    for (Map.Entry<MetricName, CellT> cell : cells.entries()) {
+      UpdateT update = cell.getValue().getDeltaUpdate(includeZero);
+      if (update != null) {
+        updates.add(MetricUpdate.create(MetricKey.create(stepName, cell.getKey()), update));
+      }
+    }
+    return updates.build();
+  }
+
+  /**
+   * Return the delta updates for all metrics in this container.
+   */
+  public MetricUpdates getDeltas() {
+    boolean includeZero = !committedFirstDelta.get();
+    return MetricUpdates.create(
+        extractUpdates(counters, includeZero),
+        extractUpdates(distributions, includeZero));
+  }
+
+  private <UpdateT, CellT extends MetricCell<UpdateT>>
+  void commitDeltas(MetricsMap<MetricName, CellT> cells, Iterable<MetricUpdate<UpdateT>> updates) {
+    for (MetricUpdate<UpdateT> counterUpdate : updates) {
+      CellT cell = checkNotNull(cells.get(counterUpdate.getKey().metricName()),
+          "Metric cell %s in delta being committed should have already been defined.");
+      cell.commitDeltaUpdate(counterUpdate.getUpdate());
+    }
+  }
+
+  /**
+   * Record all of the given deltas as reported.
+   *
+   * <p>The updates in {@code deltas} must have been produced by calling {@link #getDeltas()} on
+   * this metrics container, and no deltas should have been committed since the associated call to
+   * {@link #getDeltas()}.
+   */
+  public void commitDeltas(MetricUpdates deltas) {
+    commitDeltas(counters, deltas.counterUpdates());
```

Comment:
Going with the report-dirty-cumulatives approach since it is the simplest/most-consistent way to support the various kinds of metrics.